### PR TITLE
19 implement state machine

### PR DESCRIPTION
**Pull request by AnnaGsell**
analyser functionality, ui fragments and state machine structure build together to a more or less correct app flow

### CODE REVIEW

**Review by diehmlukas — APPROVED** [review]

File: app/src/androidTest/java/de/hsaugsburg/teamulster/sohappy/HaarCascadeFaceDetectorTest.kt [hunk 1]
```diff
@@ -14,7 +14,7 @@ import junit.framework.Assert.assertTrue as assertTrue
```
Comment (mac641) [inline app/src/androidTest/java/de/hsaugsburg/teamulster/sohappy/HaarCascadeFaceDetectorTest.kt:14]: Off-topic I assume - should be a separate PR

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/analyzer/ImageAnalyzer.kt [hunk 2]
```diff
@@ -18,10 +20,43 @@ import kotlin.concurrent.thread
...
+    private var imageAnalyzerState: ImageAnalyzerState = ImageAnalyzerState.NONE
+    private var stateMachine = StateMachineUtil.getStateMachine(fragment)
+
+
+    init {
+        stateMachine.addStateChangeListener { _, new ->
+            imageAnalyzerState = when (new) {
+                is WaitingForFace -> ImageAnalyzerState.FACE_DETECTION
+                is WaitingForSmile -> ImageAnalyzerState.SMILE_DETECTION
+                is Questions -> ImageAnalyzerState.CANCEL
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/analyzer/ImageAnalyzer.kt:47]: I'd suggest adding this to line 48, as it points to the same state

**Review by limkokyew — APPROVED** [review]
Looks good to me!

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/analyzer/ImageAnalyzer.kt [hunk 3]
```diff
@@ -18,10 +20,43 @@ import kotlin.concurrent.thread
...
+                is NoSmile, is Start -> ImageAnalyzerState.CANCEL
+                else -> imageAnalyzerState
+            }
+        }
+        imageAnalyzerState = when (stateMachine.getCurrentMachineState()) {
+            is WaitingForFace -> ImageAnalyzerState.FACE_DETECTION
+            is WaitingForSmile -> ImageAnalyzerState.SMILE_DETECTION
+            is Questions -> ImageAnalyzerState.CANCEL
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/analyzer/ImageAnalyzer.kt:55]: Same as in line 47/48

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/fragment/CameraFragment.kt [hunk 4]
```diff
@@ -60,7 +60,7 @@ class CameraFragment: Fragment() {
 
         queue = BitmapQueue()
         imageAnalyzer = ImageAnalyzer(
-            this, requireActivity(), ImageAnalyzerConfig(
+            this, ImageAnalyzerConfig(
                 "de.hsaugsburg.teamulster.sohappy.analyzer.detector.facedetectorimpl.HaarCascadeFaceDetector",
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/fragment/CameraFragment.kt:64]: We need to remember to use ConfigManager in the final version -> TODO is mandatory imo

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/fragment/CameraFragment.kt [hunk 5]
```diff
@@ -132,7 +131,7 @@ class CameraFragment: Fragment() {
         requestCode: Int, permissions: Array<String?>, grantResults: IntArray
     ) {
         if (requestCode == REQUEST_CODE_PERMISSIONS) {
-            startCameraIfReady()
+            //startCameraIfReady()
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/fragment/CameraFragment.kt:134]: I'd suggest either removing this comment or replacing it with actual code 😁

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/TakeABreath.kt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * state where the user is invited to take a breath.
+ */
+class TakeABreath : State {
+    override fun consumeAction(action: Action): State {
+        return when (action) {
+            is Action.TakeABreathTimer -> Stimulus()
+            else -> {
+                Log.d("Invalid action: ", action.toString())
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/TakeABreath.kt:14]: Same as in Stimulus

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/Stimulus.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * state where a stimulus is shown to initialize a users smile.
+ */
+class Stimulus : State {
+    override fun consumeAction(action: Action): State {
+        return when (action) {
+            is Action.StimulusTimer -> WaitingForSmile()
+            else -> {
+                Log.d("Invalid action: ", action.toString())
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/Stimulus.kt:14]: Change log to "Ignored action" as it would otherwise suggest having an error in process etc

File: app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/WaitingForFace.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * state where the camera image are analyser for a face.
+ */
+class WaitingForFace : State {
+
+    override fun consumeAction(action: Action): State {
+        return when (action) {
+            is Action.WaitingForFaceTimer -> Start()
+            is Action.FaceDetected -> TakeABreath()
+            else -> {
+                Log.d("Invalid action: ", action.toString())
```
Comment (mac641) [inline app/src/main/java/de/hsaugsburg/teamulster/sohappy/stateMachine/states/WaitingForFace.kt:16]: Same as in Stimulus

**Comment by mac641** [comment]
included in this PR #34
